File: backend/controllers/authController.ts
import User from '../models/User.js';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { Request, Response } from 'express';

export const login = async (req: Request, res: Response) => {
  try {
    const { username, password } = req.body;

    // check user exists
    const user = await User.findOne({ username }).select('+password');
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }

    // password check
    const isMatch = await bcrypt.compare(password, user.password);
    if (!isMatch) {
      return res.status(400).json({ message: "Invalid password" });
    }

    // auth
    const token = jwt.sign(
      { 
        id: user._id, 
        username: user.username 
      },
      process.env.JWT_SECRET || 'fallback_secret', // Add JWT_SECRET to your .env file
      { 
        expiresIn: '1d' 
      }
    );


    res.status(200).json({
      token,
      user: {
        id: user._id,
        username: user.username,
        name: user.name
        // email: user.email
      }
    });

  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
};

File: backend/src/main/java/com/campfire/backend/controllers/AuthController.java
package com.campfire.backend.controllers;

import com.campfire.backend.models.User;
import com.campfire.backend.repositories.UserRepository;
import com.campfire.backend.security.JwtUtil;
import lombok.Data;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private JwtUtil jwtUtil;

    // We don't have a bean for BCrypt yet, so just instantiate it
    private BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();

    @Data
    public static class LoginRequest {
        private String username;
        private String password;
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody LoginRequest request) {
        try {
            User user = userRepository.findByUsername(request.getUsername()).orElse(null);
            if (user == null) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "User not found"));
            }

            if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("message", "Invalid password"));
            }

            String token = jwtUtil.generateToken(user.getUsername(), user.getId());

            Map<String, Object> response = new HashMap<>();
            response.put("token", token);
            Map<String, String> userMap = new HashMap<>();
            userMap.put("id", user.getId());
            userMap.put("username", user.getUsername());
            userMap.put("name", user.getName());
            response.put("user", userMap);

            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of("error", e.getMessage()));
        }
    }
}
